### ISSUE TITLE
Breaking changes are not thoroughly documented

### ISSUE THREAD

**Issue by datumbox** [fchollet/keras#8368]
The Keras v2.0.9 is out and it packs lots of cool features. Go team! :)

One particular challenge that we face every time there is a new release is the fact that breaking changes are not documented thoroughly. As a result we are forced to scan through all the commit messages to make sure they will not break our code. For instance in this release we spotted a couple of breaking changes which are not included in the release notes:
- The `num_class` variables are renamed to `num_classes`
- The `random_seed` variables are renamed to `seed`
- The `Image.Iterator` inherits directly from `Sequence`, which means any class that extended it is now considered a Sequence. An additional update on the `fit_generator()` leads to the use of `OrderedEnqueuer` instead of the `GeneratorEnqueuer` if the underlying generator is a sequence which can break your code.

The fact that Python does not have proper encapsulation complicates the matters, as it relies on conventions to mark private members. In addition there is currently no way to mark the classes that should be considered public vs developer_api. 

I think it would be great within the next releases, to mark all the "private" class methods/vars with "_" if they are meant for internal use, create an annotation to distinguish between public and "private" classes and ensure that whenever we change public parts of the code to include them on the release notes.

Any thoughts? :)

cc @fchollet @farizrahman4u @ozabluda @taehoonlee @Dref360

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by ozabluda** [comment]
Private attributes and methods are already marked with "_". I don't think I ever saw a documented attribute anywhere in Keras, so all of them are undocumented - i.e. not a part of an API, although gratuitous changes to them is avoided.

**Comment by datumbox** [comment]
Hey Oleg. :)

So to paraphrase , we currently modify non-private attributes and methods without notifying the community because any undocumented class/method/attribute is considered internal, right?

**Comment by ozabluda** [comment]
You can see the discussion of the API breakage in PR https://github.com/fchollet/keras/pull/8120 and  https://github.com/fchollet/keras/pull/8235 (Issue https://github.com/fchollet/keras/issues/8227)
